## Operando con variables

Hemos visto ya diferentes tipos de estructuras de datos y como podemos almacenarlos bajo unos seudonimos denominados variables. La pregunta que puede surgir ahora es: ¿Puedo hacer algo mas con estas variables?
Es el momento de introducir a las funciones.

Una función en programacion es un subprograma o un subalgoritmo el cual opera con unos parametros de entrada. Esta nos permitirá resolver una tarea especifica. En Python ya existen unas funciones prestablecidas, pero de igual forma podemos crear unas nuevas para resolver problemas muy particulares. 
**A continuacion se mostraran algunas funciones u operaciones que ya existen y son muy utiles para el manejo de variables**

Existen distintos tipos de operadores o funciones a la hora del manejo de variables. La forma mas sencilla de acceder a ellos es escribiendo el nombre de la variable, seguido de un punto y seguido a esto pulsar la tecla TAB `nombre_de_la_variable.<TAB>`. Se desplegara una lista de funciones pre-definidas dependiendo del tipo de variable que estemos tratando, por ejemplo, en la imagen de acontinuacion se muestra la lista de funciones disponible para una variable tipo cadena o String.


![](Desktop/MOCs/cadena.PNG)


Ya que vimos un ejemplo del listado de operadores que ofrece Python para las variables tipo String, iniciemos con ellas. 

### Operando con Strings

Iniciaremos con dos operaciones muy sencillas que se pueden aplicar a las variables String, estas son .Upper() y .Lower(). Estas se aplican a la variable pero no la modifican como tal, solo nos muestran una version alterada de esta.




In [52]:
mi_string =("¡2¡Hablemos de funciones !1!")
mi_string.upper()

'¡2¡HABLEMOS DE FUNCIONES !1!'

In [53]:
mi_string.lower()

'¡2¡hablemos de funciones !1!'

In [54]:
mi_string


'¡2¡Hablemos de funciones !1!'

Notas
   * Se muestra una version modificada de la variable pero esta no se modifica.
   * No importa que la cadena de caracteres tenga numeros, estos simplemente no se afectan por la función.

De igual forma, podemos contar cuantas veces esta cierto caracter o palabra en un String, para esto utilizamos la funcion .count("parametro de busqueda") y le damos como parametro el caracter o la palabra que deseamos buscar. Un ejemplo:

In [55]:
nuevo_string= ("La casa se llenó de amor. Aureliano lo expresó en versos que no tenían principio ni fin. Los escribía en los ásperos pergaminos que le regalaba Melquíades, en las paredes del baño, en la piel de sus brazos, y en todos aparecía Remedios transfigurada: Remedios en el aire soporífero de las dos de la tarde, Remedios en la callada respiración de las rosas, Remedios en la clepsidra secreta de las polillas, Remedios en el vapor del pan al amanecer, Remedios en todas partes y Remedios para siempre")
nuevo_string

'La casa se llenó de amor. Aureliano lo expresó en versos que no tenían principio ni fin. Los escribía en los ásperos pergaminos que le regalaba Melquíades, en las paredes del baño, en la piel de sus brazos, y en todos aparecía Remedios transfigurada: Remedios en el aire soporífero de las dos de la tarde, Remedios en la callada respiración de las rosas, Remedios en la clepsidra secreta de las polillas, Remedios en el vapor del pan al amanecer, Remedios en todas partes y Remedios para siempre'

In [56]:
#Queremos saber cuantas veces utiliza Gabriel Garcia la palabra "Remedios" en este parrafo.
nuevo_string.count("Remedios")

7

Pero si lo que queremos es saber si se utilizo una palabra en particular, como puede ser "soporífero", existe una funcion .find() la cual nos indica la posicion de la palabra en cuestion. vamos con el ejemplo.

In [57]:
nuevo_string.find("soporífero")

271

Este 271 hace referencia a la posicion de esta palabra en el parrafo ?Qué pasaría si buscamos entonces una palabra que no existe en el parrafo?

In [58]:
# ?Hablaran de sancho en el texto?
nuevo_string.find("Sancho")

-1

Obtendremos un valor de `-1`.

Volvamos con nuestra oracion inicial ?se dieron cuenta que tenía un `1` entre los signos de exclamación? pues existe una forma de corregir esto. la funcion .replace("a reemplazar", "simbolo nuevo") tiene como parametro el caracter que queremos reemplazar en un String y el valor por el cual lo vamos a reemplazar. corrijamos nuestro error.

In [59]:
mi_string.replace("1","!")
mi_string.replace("2","¡")

'¡¡¡Hablemos de funciones !1!'

**Nota: Se reemplazarán todos los valores puestos en el primer parametro por los dados en el segundo**

In [36]:
ejemplo= ("111111111")
ejemplo.replace("1","hola")

'holaholaholaholaholaholaholaholahola'

En la tabla a continuación se mostrarán las funciones que aprendimos en esta seccion y la descripcion de las mismas

|Función|Descripción|
|-------|-----------|
|.Upper()| Modifica temporalmente la cadena para que sus caracteres esten en mayuscula.|
|.Lower()| Similar a la función .Upper(), hace que sus caracteres estén en minuscula.|
|.Count("x")| Cuenta el numero de "x" que hayan en la cadena.|
|.Find("x")| Da la posición de "x" en la cadena, retorna "-1" si no se encuentra.|
|.Replace("x","y")| Cambia todos los valores "x" en la cadena por valores "y"|

### Operando con Listas

En cuanto a las listas, existen funciones las cuales permiten operar sobre ellas de forma rapida y sencilla. En este apartado vamos a ver algunas funciones muy utiles que son aplicables sobre estas. 

En caso de tener una lista con valores numericos, existe la posibilidad de saber su longitud, sumar los valores de la lista, valor maximo y minimo, sin necesidad de emplear procedimientos muy complicados. para esto podemos emplear las siguientes funciones:

   * Len(lista)
   * Max(lista)
   * Min(lista)
   * sum(lista)



In [43]:
lista_numerica = [1,2,3,4,5,6,7,8,9,10]
#La forma de emplear las funciones Max, Min, Len y Sum, es la siguiente
print("Minimo:", min(lista_numerica))
print("Maximo:", max(lista_numerica))
print("longitud:",len(lista_numerica))
print("la suma de los valores de la lista es:", sum(lista_numerica))


Minimo: 1
Maximo: 10
longitud: 10
la suma de los valores de la lista es: 55


La forma de emplear las funciones anteriores, como pudimos observar, es un poco diferente a las que vimos con anterioridad. En este caso, se escribe primero el nombre de la función y entre parentesis el nombre de la lista a operar.

Así como en los Strings podemos contar la repeticion de un caracter a lo largo del mismo, en las listas podemos contar la repeticion de un elemento de la lista (ya sea numero o una palabra). Lo hacemos con la función .count("a")



In [44]:
lista_frutas = ["mandarina", "manzana", "uvas", "banano", "fresas", "mandarina", "banano"]
lista_frutas.count("mandarina")

2

Si queremos agregar otro elemento a la lista, tenemos dos operadores disponibles, si es importante la posicion en la cual agregamos un nuevo elemento utilizaremos la funcion .Intert(posicion, elemento) pero si esto no es importante podriamos utilizar la fucion .Append(elemento a agregar).
   * .Inster(pos, elemento).
   * .Append(elemento). Este lo agrega en el ultimo lugar por defecto.

In [45]:
lista_frutas.append("coco")
lista_frutasb

['mandarina',
 'manzana',
 'uvas',
 'banano',
 'fresas',
 'mandarina',
 'banano',
 'coco']

In [51]:
#Insertaremos el aguacate en la posicion 0
lista_frutas.insert(0,"aguacate")
lista_frutas

['aguacate', 'uvas', 'banano', 'fresas', 'mandarina', 'banano', 'coco']

Si queremos remover uno, existen dos funciones diferentes. podemos utilizar la funcion .Remove(elemento a remover), si deseamos remover un elemento en particular o la funcion .pop(posicion del elemento a borrar), para borrar el elemento de una posicion en particular.
   * .Pop(posicion). Es importante recordar que la posicion 1 en Python es la posicion 0
   * .Remove(elemento a remover)

In [46]:
lista_frutas.remove("mandarina")
lista_frutas

['manzana', 'uvas', 'banano', 'fresas', 'mandarina', 'banano', 'coco']

In [49]:
lista_frutas.pop(0)
lista_frutas

['uvas', 'banano', 'fresas', 'mandarina', 'banano', 'coco']

Como podemos observar, solo remueve el primer elemento que coincida con el parametro para remover.

Volviendo al ejemplo numerico, podemos ordenar la lista en el sentido contrario al que se encuentra y ordenarla de menor a mayor, para esto utilizamos las siguiente funciones:
   * .Reverse()
   * .Sort()

In [47]:
lista_numerica.reverse()
lista_numerica

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

In [48]:
lista_numerica.sort()
lista_numerica

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

**Resumen de las funciones, quiza con alguna otra mas, y su descripcion.**


|Función|Descripción|
|-------|-----------|
|.sorted() o sort()|Esta funcion sirve para ordenar una lista (numerica) ascendentemente|
|.append()| Agrega un elemento al final de la lista|
|.insert(pos , a) | inserta un valor "a" en la posicion "pos"|
|.extend(otra_lista)| Concatena la lista actual con otra_lista|
|.remove(a)| Remueve de la lista un valor que sea igual a "a" (el primero que encuentre por si hay mas de uno)|
|.pop(pos)| Remueve el elemento de la posicion "pos" de la lista|
|.index(a)| Devuelve el indice o posición de el primer elemento igual a "a" en la lista|
|.count(a)| Cuenta el numero de veces que "a" está en la lista|
|.reverse()|Invierte los elementos de la lista|
|.copy()| Devuelve una copia de la lista|
|len()| Nos dice la longitud de la lista (o tupla)|
|max()| Devuelve el maximo valor|
|min()| Devuelve el menor valor|
|sum()| Suma todos los valores de la lista|